### ISSUE TITLE
Help with PK names

### ISSUE THREAD

**Issue by elymatos** [atk4/data#849]
I'd like to thanks for the very good job! I'm trying to move my existing code from ORM to atk4/data but I'm having trouble with PK fields names. In my schema, PK names are the form id[TableName] and not just "id". atk4/data is prepending table name for the ID name and I have to duplicate the field definition. For example, at a "ElementModel" related to a "FrameModel" I have something like that:

```
$this->addField('idFrame', ['type' => 'integer']);
$this->addField('frame_idFrame', ['actual'=> 'idFrame','type' => 'integer']);
$this->hasOne('frame_idFrame', ['model' => FrameModel::class]);
```
As a novice to the framework I'd like to know if I'm doing something wrong :-) (although it is working)
Thanks
Ely

**Comment by DarkSide666** [comment]
Well, you simply have to do this in one line I guess:
```
$this->hasOne('frame_idFrame', [
    'model' => FrameModel::class,
    'our_field' => 'idFrame',
    'their_field' => 'id', // by default
]);
```
Probably no need to define type=>integer, should work without that even if it will think that id is string type. But you should be able to add type=>integer in hasOne too.

BTW, why not simply? :
```
$this->hasOne('idFrame', ['model' => FrameModel::class]);
```

**Comment by elymatos** [comment]
Thanks for comment! Actually that was my first try
> 
> BTW, why not simply? :
> 
> ```
> $this->hasOne('idFrame', ['model' => FrameModel::class]);
> ```

But then I get

```
atk4\data\Exception: Field is not defined in model
              model: App\Models\FrameElementModel
              field: 'frame_idFrame
```
 I don't know if documentation is updated at this topic (https://agile-data.readthedocs.io/en/develop/references.html#hasone-reference) but it says: " Also this reference will define a system new field ... if you haven’t done so already" 
But it is not happening :-( I really don't know if it is a bug or it is my fault in some point (maybe some configuration).
Thanks,
Ely

**Comment by PhilippGrashoff** [comment]
Hi,

the error message is strange as 'frame_' is prepended to the field name. Could you provide the complete code of FrameModelElement::init()?

Best regards

**Comment by elymatos** [comment]
> Could you provide the complete code of FrameModelElement::init()?

Yep! Here for FrameModel:

```
namespace App\Models;

use atk4\data\Model;

class FrameModel extends Model
{

    public $table = 'frame';
    public $id_field = 'idFrame';
    public $title_field = 'name';

    protected function init(): void
    {
        parent::init();

        $this->addField('name', ['type' => 'string']);
        $this->addField('description', ['type' => 'string']);
        $this->addField('root', ['type' => 'string']);
        $this->addField('active', ['type' => 'integer']);
        $this->addField('isDraft', ['type' => 'integer']);
        $this->addField('category', ['type' => 'string']);
        $this->addField('idEntity', ['type' => 'integer']);

        $this->hasMany('FE', ['model' => FrameElementModel::class]);
        $this->hasMany('LU', ['model' => LUModel::class]);

        $this->hasMany('FrameLanguage', ['model' => FrameLanguageModel::class]);

        $this->onHook(Model::HOOK_AFTER_SAVE, function ($m, $isUpdate) {
            $tl = new TimelineModel($m->persistence);
            $tl->insert([
                'entity' => 'FR',
                'id' => $m->data['idFrame'],
                'operation' => $isUpdate ? 'U' : 'I',
            ]);
        });

    }
}
```
Here for FrameElementModel (the lines i'm actually using are commented):

```
namespace App\Models;

use atk4\data\Model;

class FrameElementModel extends Model
{
    public $table = 'frameelement';
    public $id_field = 'idFrameElement';
    public $title_field = 'name';

    protected function init(): void
    {
        parent::init();

        $this->addField('name', ['type' => 'string']);
        $this->addField('description', ['type' => 'string']);
        $this->addField('type', ['actual' => 'coreType', 'type' => 'string']);
        $this->addField('active', ['type' => 'integer']);
        $this->addField('isSelected', ['type' => 'integer']);
        $this->addField('idEntity', ['type' => 'integer']);
        $this->addField('idColor', ['type' => 'integer']);

        //$this->addField('idFrame', ['type' => 'integer']);
        //$this->addField('frame_idFrame', ['actual'=> 'idFrame','type' => 'integer']);
        //$this->hasOne('frame_idFrame', ['model' => FrameModel::class]);

        $this->hasOne('idFrame', ['model' => FrameModel::class]);

        $this->onHook(Model::HOOK_AFTER_SAVE, function ($m, $isUpdate) {
            $tl = new TimelineModel($m->persistence);
            $tl->insert([
                'entity' => 'FE',
                'id' => $m->data['idFrameElement'],
                'operation' => $isUpdate ? 'U' : 'I',
            ]);
        });

    }
}

```

Thanks!
Ely

**Comment by PhilippGrashoff** [comment]
Very strange that this code causes this error message. When you comment the hasOne() line the error disappears?

**Comment by elymatos** [comment]
> Very strange that this code causes this error message. When you comment the hasOne() line the error disappears?

Wow, the error remains! So it is not related to 'hasOne'. As I have suspected from beginning, maybe it is my fault in some point...This is the code using these models:

```
                $frame = new FrameModel($db);
                $frame->load($this->data->id);
                foreach ($frame->ref('FE') as $row) {
                    $fe = (object)$row->data;
                ....
```
Maybe the error is at "hasMany" at FrameModel:

`$this->hasMany('FE', ['model' => FrameElementModel::class]);`

The framework might be trying to infer the name of FK...How can I define explicitly the FK name (idFrame)?
Thanks again!
Ely

**Comment by mvorisek** [comment]
You can start with https://github.com/atk4/ui/blob/develop/demos/init-db.php , all fields there has a different actual name and also different than `id` PK. There were some minor issues with edge cases which will be fixed in rel v2.4, which will be out in the next days, or you can use `dev-develop` version now.

**Comment by elymatos** [comment]
> You can start with https://github.com/atk4/ui/blob/develop/demos/init-db.php , all fields there has a different actual name and also different than `id` PK. There were some minor issues with edge cases which will be fixed in rel v2.4, which will be out in the next days, or you can use `dev-develop` version now.

Ok! I found it. I think it working correctly now, using "their_field".
Thanks all for help!
Ely
